### ISSUE TITLE
Ollama running locally with very high latency

### ISSUE THREAD

**Issue by vsatyakiran** [ollama/ollama#4843]
### What is the issue?

I have installed ollama and tried to run llama2, llama3:8b but it is generating just 5 to 8 tokens per second , my system config: windows os, 16gb RAM
I also tried it on ec2 instance  in aws with g5.xlarge instance type but facing the same latency, why it is happening?

### OS

Windows

### GPU

Intel

### CPU

Intel

### Ollama version

0.1.39

**Comment by Fe1777** [comment]
Certainly! Here's a GitHub Issue in the proper format for reporting the high latency experienced while running `ollama` locally and on AWS EC2 instances.

```markdown
# Issue #4843 - High Latency with Ollama Running Locally and on AWS EC2

## Description of the Issue

I have installed `ollama` and tried to run `Ilama2` and `Ilama3:8b`. However, the performance is significantly lower than expected, generating only 5 to 8 tokens per second. This issue is observed both on my local system and on an AWS EC2 instance.

### Local System Configuration:
- **Operating System:** Windows
- **RAM:** 16GB
- **CPU:** Intel (Specific model not mentioned)
- **GPU:** Intel

### AWS EC2 Configuration:
- **Instance Type:** g5.xlarge

Despite the hardware differences, both environments exhibit the same high latency.

---

### Steps to Reproduce

1. Install `ollama` version `0.1.39`.
2. Attempt to run `Ilama2` or `Ilama3:8b`.
3. Monitor the token generation rate, which is consistently between 5 to 8 tokens per second.

### Actual Result

In both local and EC2 environments, the application generates only 5 to 8 tokens per second, indicating high latency.

### Expected Result

A higher rate of token generation that aligns with the capabilities of the specified hardware, especially on a powerful g5.xlarge EC2 instance.

---

### Environment Details

**Local System:**
- **OS:** Windows
- **RAM:** 16GB
- **CPU:** Intel

**AWS EC2 Instance:**
- **Instance Type:** g5.xlarge

**Ollama Version:**
- **0.1.39**

---

### Additional Information

- **EC2 Instance type `g5.xlarge`** is typically optimized for GPU-based workloads.
- **Local System** resources (16GB RAM and an Intel CPU) should provide better performance than experienced.
- The issue is consistent across different environments indicating a potential software issue or configuration problem.

---

### Screenshots / Logs

*(Please attach any relevant logs or screenshots that can help in diagnosing the issue.)*

---

### Tagging

**Reported by:** vsatyakiran

**Maintainer:** *(Tag relevant maintainer if known)*

### Comments

Any insights, suggestions, or guidelines to improve performance would be greatly appreciated. Thank you!

---

End of Issue
```

This issue report includes a detailed description of the problem, the steps to reproduce it, expected vs. actual results, environmental details, and a request for assistance. It ensures that sufficient information is provided to help diagnose and address the issue.

**Comment by pdevine** [comment]
@vsatyakiran you're almost certainly running on the CPU instead of the GPU. What is the output of `ollama ps`? Also, what type of GPU are you trying to use? With the g5.xlarge I believe it's an a100 and there was an issue w/ nvidia driver version 555 which should be fixed in Ollama 0.1.40.

**Comment by Fe1777** [comment]
It sounds like you're dealing with a performance issue where your computations are running on the CPU rather than the GPU. Let's address your problem step by step.

**1. Check if Processes are Running on GPU:**
First, you need to determine whether your processes are indeed running on the GPU. To do that, you can run the command mentioned:

```sh
ollama ps
```

This command should give you an overview of the processes and indicate whether they're utilizing the GPU or not.

**2. Inspect Your GPU Type:**
You mentioned you're working with a `g5.xlarge` instance. Indeed, this instance type typically comes with an NVIDIA A100 GPU. 

**3. Check for Driver Issues:**
There may be an issue with the NVIDIA driver version you are using. Specifically, version 555 has known issues with the A100 GPU. Make sure you have the correct driver version installed for your GPU.

**4. Update Ollama:**
Ensure you have the latest version of Ollama installed, as the issue with driver version 555 was purportedly fixed in Ollama version 0.1.40. You can update Ollama to the latest version using the package manager you've installed it with, typically:

```sh
pip install --upgrade ollama
```

or if you're using `conda`:

```sh
conda update ollama
```

**5. Verify NVIDIA Drivers:**
Check the NVIDIA driver version on your machine. You can do this using the following command:

```sh
nvidia-smi
```

This will display the details of the installed NVIDIA driver and should help you confirm which version you are running. If you find it’s version 555, you may need to update the drivers to a more suitable version.

**6. Transition Database to GPU:**
Finally, ensure that your workload configurations are set to utilize the GPU. Some libraries and frameworks require explicit instructions to switch computations to the GPU.

For example, in PyTorch, you can transfer a tensor to the GPU using:

```python
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
tensor = tensor.to(device)
```

In TensorFlow, you might set the GPU with:

```python
import tensorflow as tf

gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    try:
        # Currently, memory growth needs to be the same across GPUs
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
    except RuntimeError as e:
        # Memory growth must be set before GPUs have been initialized
        print(e)
```

**Summary:**
- Confirm processes are running on GPU with `ollama ps`.
- Ensure you’re using the correct GPU and that it's not disabled.
- Update Ollama to version 0.1.40 or later.
- Verify and, if necessary, update the NVIDIA driver.
- Make sure your workflow explicitly targets the GPU.

By following these steps, you should be able to identify and resolve the issue, ensuring your computations leverage the GPU’s performance benefits.

**Comment by dhiltgen** [comment]
@vsatyakiran if you're still having troubles, please upgrade to the latest version, and if that doesn't get you running on your GPU, please share your server log and I'll reopen the issue.

https://github.com/ollama/ollama/blob/main/docs/troubleshooting.md
